### PR TITLE
(with error to look into further) validation rule 184

### PR DESCRIPTION
**Pull request by jfan2006**
Trying to do the rule 184 #105 but still can't figure out how to do it... 
(saying error in line 24 of validators.py)
Could you help me a bit more on this please? Thanks

### CODE REVIEW

File: validator903/validators.py [hunk 1]
```diff
@@ -1,6 +1,33 @@
...
+            child_record = dfs['Header']
+            ad1 = dfs['AD1']
+
+            all_data = ad1.reset_index().merge(child_record, how='left', on='CHILD')
+
+            mask = (all_data['DATE_PLACED'] >= all_data['DOB']) | all_data['DATE_PLACED'].isna()
+
+            validation_error = ~mask
+            
+            validation_error_locations = ad1.index[validation_error]['index'].unique()
```
Comment (mwaddoups) [inline validator903/validators.py]: I think you just don't need the `ad1.index` part - the `'index'` column is just a regular old column, so 

```
ad1[validation_error]['index'].unique()
```

should do the trick!

**Comment by mwaddoups** [comment]
I added a comment on line 24 - take a look and see if that helps.

File: validator903/validators.py (hunk 1 above)
Comment (jfan2006) [reply validator903/validators.py]: I have just tried as you suggested but it still has errors.

Comment (mwaddoups) [reply validator903/validators.py]: Are you able to provide some more details - like screenshots of the error or something? That might be helpful in diagnosing.

Comment (dezog) [reply validator903/validators.py:24]: Hi Fan, are you sure the DataFrame has a column called `'index'`? 
Maybe try keeping the `.index` and dropping the `['index']`:
```
ad1.index[validation_error].unique()
```

Comment (dezog) [reply validator903/validators.py]: Looks like `all_data` will have a column called `'index'` due to  the `reset_index()` in line 18, but `ad1` won't have had its index reset. So maybe instead try:
```
ad1.index[validation_error].unique()
```
That make sense, @mwaddoups?

Comment (mwaddoups) [reply validator903/validators.py]: Now I'm confused! I think it needs to be `all_data` to read the appropriate index values off.

```
all_data[validation_error]['index'].unique()
```

**Review by dezog — CHANGES_REQUESTED** [review]
Just setting this to show 'changes requested'. 

Let me know if the edit Mark suggested in https://github.com/SocialFinanceDigitalLabs/quality-lac-data-beta-validator/pull/314#discussion_r700194448 doesn't do the trick, but it looks like it should do. 🤞

**Comment by jfan2006** [comment]
Tried and made changes based on Mark's suggestion above, and also dates comparison. 
But this still has errors....

File: tests/test_validators.py [hunk 2]
```diff
@@ -1,6 +1,23 @@
...
+    fake_data_header = pd.DataFrame({
+        'CHILD': ['111', '112', '113', '114'], 
+        'DOB': ['01/10/2017', '31/05/2018', '10/03/2019', '19/08/2020'],
+    })
+    fake_data_ad1 = pd.DataFrame({
+        'CHILD': ['111', '112', '113', '114'], 
+        'DATE_PLACED': ['01/10/2017', '10/02/2019', '10/02/2019', pd.NA],
+    })
+
+    fake_dfs = {'Header': fake_data_header, 'AD1': fake_data_ad1}
```
Comment (dezog) [inline tests/test_validators.py]: AD1 > PlacedAdoption

File: tests/test_validators.py [hunk 3]
```diff
@@ -1,6 +1,23 @@
...
+    
+    error_defn, error_func = validate_184()
+
+    result = error_func(fake_dfs)
+
+    assert result == {'AD1': [2]}
```
Comment (dezog) [inline tests/test_validators.py]: AD1 > PlacedAdoption

File: validator903/validators.py [hunk 4]
```diff
@@ -1,6 +1,36 @@
...
 
+def validate_184():
+    error = ErrorDefinition(
+        code='184',
+        description='Date of decision that a child should be placed for adoption is before the child was born.',
+        affected_fields=['DATE_PLACED','DOB'],
+    )
+
+    def _validate(dfs):
+        if 'Header' not in dfs or 'AD1' not in dfs:
```
Comment (dezog) [inline validator903/validators.py]: AD1 > PlacedAdoption

File: validator903/validators.py [hunk 5]
```diff
@@ -1,6 +1,36 @@
...
+            all_data['DATE_PLACED'] = pd.to_datetime(all_data['DATE_PLACED'],format='%d/%m/%Y',errors='coerce')
+            all_data['DOB'] = pd.to_datetime(all_data['DOB'],format='%d/%m/%Y',errors='coerce')
+
+            mask = (all_data['DATE_PLACED'] >= all_data['DOB']) | all_data['DATE_PLACED'].isna()
+
+            validation_error = ~mask
+            
+            validation_error_locations = all_data[validation_error]['index'].unique()
+        
+            return {'AD1': validation_error_locations.tolist()}
```
Comment (dezog) [inline validator903/validators.py]: AD1 > PlacedAdoption

File: validator903/validators.py [hunk 6]
```diff
@@ -1,6 +1,36 @@
...
+        description='Date of decision that a child should be placed for adoption is before the child was born.',
+        affected_fields=['DATE_PLACED','DOB'],
+    )
+
+    def _validate(dfs):
+        if 'Header' not in dfs or 'AD1' not in dfs:
+            return {}
+        else:
+            child_record = dfs['Header']
+            ad1 = dfs['AD1']
```
Comment (dezog) [inline validator903/validators.py]: AD1 > PlacedAdoption

**Review by dezog — CHANGES_REQUESTED** [review]
Hey Fan, think I've figured this one out - it's looking for DATE_PLACED in the AD1 table, but that's actually a field of the PlacedAdoption table. 

That's also causing an error in one of the integration tests where merging AD1 & Header leads to 'DOB_x' & 'DOB_y' columns since they both have that field, so accessing DOB raises a KeyError - but switching to PlacedAdoption should fix this.

I've tried to indicate all the relevant lines to change in the code, hopefully that does it!

**Review by dezog — APPROVED** [review]

**Pull request merged**